internal/pkg/agent/application/info/state.go

```diff
@@ -21,7 +21,7 @@ const (
 // This verifies the running executable path based on hard-coded paths
 // for each platform type.
 func RunningInstalled() bool {
-	expectedPaths := []string{filepath.Join(paths.InstallPath, paths.BinaryName)}
+	expectedPaths := []string{filepath.Join(paths.Top(), paths.BinaryName)}
```

Comment:
The result of this function can change the lookup location of the control socket of the Elastic Agent. With this change I believe that it means that the control socket will always be in the expected install path even if the Elastic Agent is installed into a different path.

I wonder if this is the behavior we want in this case or would we expect it to stay nested into the directory? Because I believe this change will result in the control protocol not being used under the elastic-agent directory at all.

Reply:
Yeah, you're right - with this change, the `RunningInstalled()` function will always return `true` now and one impact of that is that Agent would always try to use the sockets (control and shipper) from their fixed locations.  I agree that this is probably not the behavior we want - in fact, I don't think a non-installed Agent would even work correctly with this change.  Let me test and try to make the necessary fixes...

Reply:
@blakerouse I'm trying to think of how to fix this situation so the result will be that for non-installed Agents we use the socket files under the Agent's install path but for installed Agents we use the fixed paths, e.g. `/var/run/elastic-agent.sock`.  

I have an idea I want your opinion on, because it seems too easy/simple 🙂.  What if we make the `./elastic-agent instal` code path write a `.installed` file in the Agent's topPath folder.  Then, this `RunningInstalled()` function could simply check for that file's existence to decide if the executing Agent is an installed Agent or not.  WDYT?

Reply:
@blakerouse I implemented the installation marker file idea from my previous comment in 810560ad3.  And I've tested it with both the installed and non-installed Agent cases.  Everything seems to work as expected so I'd be curious to get your opinion on the idea and the implementation - can you think of anything that might not work or break because of this approach?

Reply:
I am not a big fan of this approach as it relies on a file to be present, if that file is lost for any reason we have an Elastic Agent that now works differently.

Could we change the `RunningInstalled` back to how you had it and then add a new one like `RunningDefaultInstalled` that checks if the Elastic Agent is installed in the default base path. Then use the `RunningDefaultInstalled` in the `pkg/control/addr*.go` and the `pkg/component/runtime/manager_shipper_*.go`. That way current installation directories do not change, but if you use a custom base path they stay nested in that base path.

Reply:
So, just so we're on the same page, this is what you're proposing (in terms of the effects of the change):

| Method of installation / execution             | Control socket path (MacOS) | Shipper socket path (MacOS) |
| ---------------------------------------- | ----------------------------- | ------------------------------ |
| `./elastic-agent install`                               | `/var/run/elastic-agent.sock` | `/var/run/elastic-agent-{component-id}-pipe.sock` |
| `./elastic-agent install --base-path /foo`   | `/foo/Elastic/Agent/data/tmp/elastic-agent-control.sock` | `/foo/Elastic/Agent/data/tmp/elastic-agent-{component-id}-pipe.sock` |
| `./elastic-agent run`   | `./data/tmp/elastic-agent-control.sock` | `./data/tmp/elastic-agent-{component-id}-pipe.sock` |

I have a couple of questions about this approach:

1. If it's okay for installed Agents to have two different paths for their control and shipper sockets, depending on whether `--base-path` was used or not, why not always keep the sockets under the `{topPath}/data/tmp/`?  Why do we keep sockets in `/var/run` for the default base path installation case - what is this buying us?
   1. If the answer is that keeping sockets under `/var/run` does buy us something more, then we would now lose that with the custom base path installation scenario.  I think this makes for an odd user experience because as far as they're concerned they're "just" changing the base path of the installation.
   2. If the answer is that keeping sockets under `/var/run` doesn't buy us anything more, then why not simplify and always store sockets under `{topPath}/data/tmp/`?

Reply:
Just so we're on the same page, this is what you're proposing (in terms of the effects of the change):

| Method of installation / execution             | Control socket path (MacOS) | Shipper socket path (MacOS) |
| ---------------------------------------- | ----------------------------- | ------------------------------ |
| `./elastic-agent install`                               | `/var/run/elastic-agent.sock` | `/var/run/elastic-agent-{component-id}-pipe.sock` |
| `./elastic-agent install --base-path /foo`   | `/foo/Elastic/Agent/data/tmp/elastic-agent-control.sock` | `/foo/Elastic/Agent/data/tmp/elastic-agent-{component-id}-pipe.sock` |
| `./elastic-agent run`   | `./data/tmp/elastic-agent-control.sock` | `./data/tmp/elastic-agent-{component-id}-pipe.sock` |

I have a couple of questions about this approach:

1. If it's okay for installed Agents to have two different paths for their control and shipper sockets, depending on whether `--base-path` was used or not, why not always keep the sockets under the `{topPath}/data/tmp/`?  Why do we keep sockets in `/var/run` for the default base path installation case - what is this buying us?
   1. If the answer is that keeping sockets under `/var/run` does buy us something more, then we would now lose that with the custom base path installation scenario.  I think this makes for an odd user experience because as far as they're concerned they're "just" changing the base path of the installation.
   2. If the answer is that keeping sockets under `/var/run` doesn't buy us anything more, then why not simplify and always store sockets under `{topPath}/data/tmp/`?

> I am not a big fan of this approach as it relies on a file to be present, if that file is lost for any reason we have an Elastic Agent that now works differently.

As for this concern, it's a fair one, but I don't see how how/why we'd lose this file.  We can write a test (probably integration or E2E) that can ensure that this file is always present for the `./elastic-agent install` scenario.  And if a user deletes this file, to me that's messing with internals so all bets are off at that point from a supportability perspective.  Maybe we could include something in the output of `elastic-agent status` and/or `elastic-agent diagnostics` to indicate the presence/absence of the file?

I especially feel it's worth pursuing the installation marker file approach if it gives a more consistent user experience (see 1.1 above).

Reply:
You captured the results in the table perfectly.

Keeping the paths under the `/var/run` allows for any other tool on the system to be able to interact with the Elastic Agent knows that it is an installed Elastic Agent. That is the only thing I see we get out of it, which is very little. I would almost say we could just switch it all to `{topPath}/data/tmp/` but that would be a breaking change, so I don't want to just make that change.

Really the only reason I bring this up is because I don't like the `.installed` file at all as you noted we would need to add that information to `status` and `diagnostics` for supportability and I would prefer to remove the need for that. I don't want to get SDH's and have to figure out that they deleted the `.installed` file and that is the reason something is not working.

If you revert to what you had back in `RunningInstalled` would it result in the paths of the sockets always being placed in `/var/run` even if installed with a different `--base-path`? I honestly would prefer that more, than using the `.installed` file.

Reply:
> If you revert to what you had back in `RunningInstalled` would it result in the paths of the sockets always being placed in `/var/run` even if installed with a different `--base-path`? I honestly would prefer that more, than using the `.installed` file.

Yes, if I revert all the way back to the implementation that started this conversation: https://github.com/elastic/elastic-agent/pull/2500#discussion_r1172701236, the the socket would always be under `/var/run/`.  The only "issue" I see with this implementation is what you pointed out in https://github.com/elastic/elastic-agent/pull/2500#discussion_r1172701236: that `RunningInstalled()` will _always_ return true.  So we could just delete the `RunningInstalled` function at that point, no?

Reply:
That will completely break the ability for two Elastic Agent's to be started at the same time. Imagine you have an installed Elastic Agent and then you want to run one in another directory to test something, then this PR merged with the `.installed` marker removed would break.

I think we either go with my suggestion of the `RunningDefaultInstalled` or we use the `.installed` marker you implemented, I don't see another path at the moment. I will let you make that call, I prefer the `RunningDefaultInstalled` as it doesn't require an extra file and less issues I feel like with supporting it. But if we want installation to use the `/var/run` paths for an installed Elastic Agent at a different base path, then I think the `.installed` file is the only path forward. But I don't think we should break the use-case of allowing multiple Elastic Agent's to run at the same time.

Reply:
After I submitted that comment, I did think of another possible solution. We could remove the `RunningInstalled()` as you suggested, adding a new flag for `--control-sock` as a global flag that would allow the first Elastic Agent to run with the default path. Then if you want to run a second Elastic Agent you could use the `--control-sock` parameter to place it in a different location.

I am kinda leaning toward that one now.

Reply:
> adding a new flag for `--control-sock` as a global flag that would allow the first Elastic Agent to run with the default path. Then if you want to run a second Elastic Agent you could use the `--control-sock` parameter to place it in a different location.

I want to make sure I understand this from a user's perspective.  Wouldn't the second Agent's user have to keep specifying `--control-sock` with every `elastic-agent ...` command they run that's intended for the second Agent?  That seems like an annoying UX to me, personally.   I get that running two Agents on the same host is not a common use case but I'd suggest that a user messing with the internal files of an Agent's installation (i.e. files under `{topPath}`) is even rarer.

Also, from a testing perspective, we would have to test four scenarios: the combinations of using/not using `--base-path` with `elastic-agent install` and using/not using `--control-sock`.  That's two more than than if we just go with the `--base-path` option with the `.installed` implementation.

As for the `.installed` file being removed, I'm not sure why we're treating this one as special compared to any of the _other_ files under `{topPath}`.  If a user accidentally or intentionally removed _any_ file under `{topPath}` wouldn't we be in "unexpected/unsupported behavior" territory anyway?  I think the unlikely scenario of a user accidentally deleting this file (or any other internal file for that matter) is worth the tradeoff of a simpler user experience.  And, as I mentioned in https://github.com/elastic/elastic-agent/pull/2500#discussion_r1173890045, we could mitigate this scenario by including information about the existence of this file in `elastic-agent diagnostics` and/or `elastic-agent status`.

Reply:
You are correct on the usage of `--control-sock`, they would have to specify it for every command.

Something about the `.installed` feels wrong, I just can't come up with a reason. ;-)

I am okay with using the `.installed` file.